codegen/value.go

```diff
@@ -52,9 +54,71 @@ func NewRegister(ctx context.Context) Register {
 	}
 }
 
-func (r *register) Set(iface interface{}) (err error) {
-	r.Value, err = r.ctor(iface)
-	return err
+func (r *register) Set(iface interface{}) error {
+	return r.SetAsync(func(Value) (Value, error) {
+		return r.ctor(iface)
+	})
+}
+
+func (r *register) SetAsync(f func(Value) (Value, error)) error {
+	var prev chan Value
+	if r.last == nil {
+		prev = make(chan Value, 1)
+		prev <- r.Value
+	} else {
+		prev = r.last
+	}
+
+	ret := make(chan Value)
+	r.last = ret
+
+	go func() {
+		defer close(prev)
+		next, err := f(<-prev)
+		if err != nil {
+			next = &errorValue{err}
+		}
+		ret <- next
+	}()
+
+	return nil
+}
+
+func (r *register) value() Value {
+	if r.last != nil {
+		r.Value = <-r.last
+		close(r.last)
```

Comment:
Same question here. You're done with `r.last`, so closing it doesn't change anything.